{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 配置环境\n",
    "使用HiC-Pro中的yml文件创建conda环境\n",
    "```\n",
    "mamba env create -f environment.yml\n",
    "```\n",
    "\n",
    "接下来按照github readme文件make configure和make install\n",
    "\n",
    "激活环境\n",
    "```\n",
    "conda activate HiC-Pro_v3.1.0\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 酶切图谱\n",
    "```\n",
    "python /lustre1/chengqiyi_pkuhpc/zhaohn/0.apps/HiC-Pro_installed/HiC-Pro_3.1.0/bin/utils/digest_genome.py \\\n",
    "    -r hindiii \\\n",
    "    -o hg38_hindiii.bed \\\n",
    "    /lustre1/chengqiyi_pkuhpc/zhaohn/1.database/db_genomes/genome_fa/genome_ucsc_hg38/genome_ucsc_hg38.fa\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 染色体长度\n",
    "fasta的fai文件中的前两列信息就是长度\n",
    "```\n",
    "samtools faidx /lustre1/chengqiyi_pkuhpc/zhaohn/1.database/db_genomes/genome_fa/genome_ucsc_hg38/genome_ucsc_hg38.fa\n",
    "\n",
    "cut -f1-2 /lustre1/chengqiyi_pkuhpc/zhaohn/1.database/db_genomes/genome_fa/genome_ucsc_hg38/genome_ucsc_hg38.fa.fai > chrom_hg38.sizes\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 下载测试文件\n",
    "```\n",
    "wget  https://zerkalo.curie.fr/partage/HiC-Pro/HiCPro_testdata.tar.gz && tar -zxvf HiCPro_testdata.tar.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 集群提交\n",
    "## 生成提交任务脚本\n",
    "```\n",
    "HiC-Pro -i test_data -o out_dir -c config-hicpro.txt -p\n",
    "```\n",
    "## 提交在北极星\n",
    "```\n",
    "sbatch HiCPro_step1_ZHN-HiC.sh\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 软件目录script下的make_slurm_script.sh需要修改\n",
    "\n",
    "【直接复制】\n",
    "\n",
    "```shell\n",
    "#!/bin/bash\n",
    "\n",
    "## HiC-Pro\n",
    "## Copyright (c) 2015 Institut Curie                               \n",
    "## Author(s): Nicolas Servant\n",
    "## Contact: nicolas.servant@curie.fr\n",
    "## This software is distributed without any guarantee under the terms of the BSD-3 licence.\n",
    "## See the LICENCE file for details\n",
    "\n",
    "##\n",
    "## Create SLURM Torque files\n",
    "##\n",
    "\n",
    "dir=$(dirname $0)\n",
    "\n",
    "usage()\n",
    "{\n",
    "    echo \"usage: $0 -c CONFIG [-s STEP]\"\n",
    "}\n",
    "\n",
    "MAKE_OPTS=\"\"\n",
    "\n",
    "while [ $# -gt 0 ]\n",
    "do\n",
    "    case \"$1\" in\n",
    "    (-c) conf_file=$2; shift;;\n",
    "\t(-s) MAKE_OPTS=$2; shift;;\n",
    "    (--) shift; break;;\n",
    "    (-*) echo \"$0: error - unrecognized option $1\" 1>&2; exit 1;;\n",
    "    (*)  suffix=$1; break;;\n",
    "    esac\n",
    "    shift\n",
    "done\n",
    "\n",
    "if [ -z \"$conf_file\" ]; then usage; exit 1; fi\n",
    "\n",
    "CONF=$conf_file . $dir/hic.inc.sh\n",
    "unset FASTQFILE\n",
    "\n",
    "## Define input files\n",
    "if [[ $MAKE_OPTS == \"\" || $MAKE_OPTS == *\"mapping\"* ]]\n",
    "then\n",
    "    inputfile=inputfiles_${JOB_NAME}.txt\n",
    "    ifq=$(get_hic_files $RAW_DIR .fq)\n",
    "    ifastq=$(get_hic_files $RAW_DIR .fastq)\n",
    "    echo -e \"$ifq\\n$ifastq\" | grep $PAIR1_EXT | sed -e \"s|$RAW_DIR||\" -e \"s|^/||\" > $inputfile\n",
    "    count=$(cat $inputfile | wc -l)\n",
    "elif [[ $MAKE_OPTS == *\"proc_hic\"* ]]\n",
    "then\n",
    "    inputfile=inputfiles_${JOB_NAME}.txt\n",
    "    get_hic_files $RAW_DIR .bam | grep $PAIR1_EXT | sed -e \"s|$RAW_DIR||\" -e \"s|^/||\" > $inputfile\n",
    "    count=$(cat $inputfile | wc -l)\n",
    "fi\n",
    "\n",
    "## Paralelle Implementation\n",
    "if [[ $MAKE_OPTS == \"\" || $MAKE_OPTS == *\"mapping\"* || $MAKE_OPTS == *\"proc_hic\"* ]]\n",
    "then\n",
    "    make_target=\"all_sub\"\n",
    "    ## Remove per sample steps\n",
    "    if [[ $MAKE_OPTS != \"\" ]]; then \n",
    "    make_target=$(echo $MAKE_OPTS | sed -e 's/,/ /g'); \n",
    "    make_target=$(echo $make_target | sed -e 's/merge_persample//g');\n",
    "    make_target=$(echo $make_target | sed -e 's/build_contact_maps//g');\n",
    "    make_target=$(echo $make_target | sed -e 's/ice_norm//g');\n",
    "        make_target=$(echo $make_target | sed -e 's/quality_checks//g');\n",
    "    fi\n",
    " \n",
    "    ## step 1 - parallel\n",
    "    torque_script=HiCPro_step1_${JOB_NAME}.sh\n",
    " \n",
    "    cat > ${torque_script} <<EOF\n",
    "#!/bin/bash\n",
    "#SBATCH -N 1\n",
    "#SBATCH -c ${N_CPU}\n",
    "#SBATCH -p ${JOB_QUEUE}\n",
    "\n",
    "#SBATCH --job-name=s1_${JOB_NAME}_HiCpro\n",
    "#SBATCH --export=ALL\n",
    "#SBATCH --no-requeue\n",
    "#SBATCH -A ${JOB_ACCOUNT}\n",
    "#SBATCH --qos=${JOB_QOS}\n",
    "EOF\n",
    "    \n",
    "    if [[ $count -gt 1 ]]; then\n",
    "\techo -e \"#SBATCH --array=1-$count\" >> ${torque_script}\n",
    "    fi\n",
    "    cat >> ${torque_script} <<EOF\n",
    "FASTQFILE=\\$SLURM_SUBMIT_DIR/$inputfile; export FASTQFILE\n",
    "make --file ${SCRIPTS}/Makefile CONFIG_FILE=${conf_file} CONFIG_SYS=${INSTALL_PATH}/config-system.txt $make_target 2>&1\n",
    "EOF\n",
    "    \n",
    "    chmod +x ${torque_script}\n",
    "\n",
    "    ## User message\n",
    "    echo \"The following command will launch the parallel workflow through $count torque jobs:\"\n",
    "    echo sbatch ${torque_script}\n",
    "fi    \n",
    "\n",
    "\n",
    "## Per sample Implementation\n",
    "if [[ $MAKE_OPTS == \"\" || $MAKE_OPTS == *\"build_contact_maps\"* || $MAKE_OPTS == *\"ice_norm\"* || $MAKE_OPTS == *\"quality_checks\"* ]]\n",
    "then\n",
    "    make_target=\"all_persample\"\n",
    "    ## Remove parallele mode\n",
    "    if [[ $MAKE_OPTS != \"\" ]]; \n",
    "    then \n",
    "\tmake_target=$(echo $MAKE_OPTS | sed -e 's/,/ /g'); \n",
    "\tmake_target=$(echo $make_target | sed -e 's/mapping//g');\n",
    "\tmake_target=$(echo $make_target | sed -e 's/proc_hic//g');\n",
    "    fi\n",
    "\n",
    "    torque_script_s2=HiCPro_step2_${JOB_NAME}.sh\n",
    "    cat > ${torque_script_s2} <<EOF\n",
    "#!/bin/bash\n",
    "\n",
    "#SBATCH -N 1\n",
    "#SBATCH -c 1\n",
    "#SBATCH -p ${JOB_QUEUE}\n",
    "\n",
    "#SBATCH --job-name=s2_${JOB_NAME}_HiCpro\n",
    "#SBATCH --export=ALL\n",
    "#SBATCH --no-requeue\n",
    "#SBATCH -A ${JOB_ACCOUNT}\n",
    "#SBATCH --qos=${JOB_QOS}\n",
    "\n",
    "cd \\$SLURM_SUBMIT_DIR\n",
    "\n",
    "make --file ${SCRIPTS}/Makefile CONFIG_FILE=${conf_file} CONFIG_SYS=${INSTALL_PATH}/config-system.txt $make_target 2>&1\n",
    "EOF\n",
    "    \n",
    "    chmod +x ${torque_script_s2}\n",
    "\n",
    "    ## User message\n",
    "    echo \"The following command will merge the processed data and run the remaining steps per sample:\"\n",
    "    echo sbatch ${torque_script_s2}\n",
    "fi\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# config file\n",
    "```shell\n",
    "# Copy and edit the configuration file ‘config-hicpro.txt’ in your local folder. \n",
    "# \n",
    "# The ‘[]’ options are optional and can be undefined.\n",
    "# 带有[]中括号的参数可以不定义，为可选参数\n",
    "# Please change the variable settings below if necessary\n",
    "#########################################################################\n",
    "## Paths and Settings  - Do not edit !\n",
    "## 输入输出文件路径，尽量不动\n",
    "#########################################################################\n",
    "TMP_DIR = tmp\n",
    "\n",
    "LOGS_DIR = logs\n",
    "\n",
    "BOWTIE2_OUTPUT_DIR = bowtie_results\n",
    "\n",
    "MAPC_OUTPUT = hic_results\n",
    "\n",
    "# Link to rawdata folder. The user usually not need to change this option\n",
    "# 尽量不该rawdata路径\n",
    "RAW_DIR = rawdata\n",
    "\n",
    "#######################################################################\n",
    "## SYSTEM AND SCHEDULER - Start Editing Here !!\n",
    "## 从这里开始编辑\n",
    "#######################################################################\n",
    "# ?文档中没有说明\n",
    "SORT_RAM = 1000M\n",
    "\n",
    "# name of the main log file\n",
    "LOGFILE = hicpro.log\n",
    "\n",
    "# 【可选参数】name of the job on the cluster\n",
    "JOB_NAME = ZHN-HiC \n",
    "\n",
    "# 【可选参数】队列指定\n",
    "JOB_ACCOUNT = chengqiyi_g1\n",
    "# 【cnlong】\n",
    "# N_CPU = 20\n",
    "# JOB_QUEUE = cn-long\n",
    "# JOB_QOS = chengqiyicnl\n",
    "\n",
    "# 【cn-short】\n",
    "# N_CPU = 20\n",
    "# JOB_QUEUE = cn-short\n",
    "# JOB_QOS = chengqiyicns\n",
    "\n",
    "# 【cn_nl】\n",
    "N_CPU = 20\n",
    "JOB_QUEUE = cn_nl\n",
    "JOB_QOS = chengqiyicnnl\n",
    "\n",
    "# 【fat4way】\n",
    "# N_CPU = 24\n",
    "# JOB_QUEUE = fat4way\n",
    "# JOB_QOS = chengqiyif4w\n",
    "\n",
    "# 【fat8way】\n",
    "# N_CPU = 64\n",
    "# JOB_QUEUE = fat8way\n",
    "# JOB_QOS = chengqiyif8w\n",
    "#########################################################################\n",
    "## Data\n",
    "#########################################################################\n",
    "# Keyword for first mate detection. Default:_R1\n",
    "PAIR1_EXT = _R1\n",
    "\n",
    "# Keywoard for seconde mate detection. Default:_R2\n",
    "PAIR2_EXT = _R2\n",
    "\n",
    "#######################################################################\n",
    "## Alignment options\n",
    "#######################################################################\n",
    "# Minimum mapping quality. \n",
    "# Reads with lower quality are discarded. \n",
    "# Default: 0\n",
    "MIN_MAPQ = 10\n",
    "\n",
    "# Path to bowtie2 indexes\n",
    "BOWTIE2_IDX_PATH = /lustre1/chengqiyi_pkuhpc/zhaohn/1.database/db_genomes/genome_fa/genome_ucsc_hg38/\n",
    "\n",
    "# bowtie2 options for mapping step1. \n",
    "# Default: –very-sensitive -L 30 –score-min L,-0.6,-0.2 –end-to-end –reorder\n",
    "BOWTIE2_GLOBAL_OPTIONS = --very-sensitive -L 30 --score-min L,-0.6,-0.2 --end-to-end --reorder\n",
    "\n",
    "# bowtie2 options for mapping step2. \n",
    "# Default: –very-sensitive -L 20 –score-min L,-0.6,-0.2 –end-to-end –reorder\n",
    "BOWTIE2_LOCAL_OPTIONS =  --very-sensitive -L 20 --score-min L,-0.6,-0.2 --end-to-end --reorder\n",
    "\n",
    "#######################################################################\n",
    "## Annotation files\n",
    "#######################################################################\n",
    "# Reference genome prefix used for genome indexes. \n",
    "# Default: hg19\n",
    "# 【需要升级为hg38，查阅文档】\n",
    "REFERENCE_GENOME = genome_ucsc_hg38.fa.bowtie2_index\n",
    "\n",
    "# Chromsome size file. \n",
    "# Loaded from the ANNOTATION folder in the HiC-Pro installation directory. \n",
    "# Default: chrom_hg19.sizes\n",
    "# 这个要放到安装文件夹下头！\n",
    "# 【需要升级为hg38，查阅文档】\n",
    "GENOME_SIZE = chrom_hg38.sizes\n",
    "\n",
    "#######################################################################\n",
    "## Allele specific analysis\n",
    "## http://nservant.github.io/HiC-Pro/AS.html#as\n",
    "#######################################################################\n",
    "# VCF file to SNPs which can be used to distinguish parental origin. See the allele specific section for more details\n",
    "# 【可选参数】\n",
    "# ALLELE_SPECIFIC_SNP = \n",
    "\n",
    "#######################################################################\n",
    "## Capture Hi-C analysis\n",
    "#######################################################################\n",
    "# BED file of target regions to focus on (mainly used for capture Hi-C data\n",
    "# 【可选参数】\n",
    "# CAPTURE_TARGET =\n",
    "REPORT_CAPTURE_REPORTER = 1\n",
    "\n",
    "#######################################################################\n",
    "## Digestion Hi-C\n",
    "#######################################################################\n",
    "# BED file with restriction fragments. \n",
    "# Full path or name of file available in the ANNOTATION folder. \n",
    "# Default: HindIII_resfrag_hg19.bed\n",
    "# 含有限制性片段的BED文件\n",
    "# 这个要放到安装文件夹下头！\n",
    "# 【需要升级为hg38，查阅文档】\n",
    "GENOME_FRAGMENT = HindIII_resfrag_hg38.bed\n",
    "\n",
    "# Ligation site sequence used for reads trimming. \n",
    "# Depends on the fill in strategy. \n",
    "# Example: AAGCTAGCTT\n",
    "# 酶的序列，例如：\n",
    "# HindIII，为AAGCTAGCTT\n",
    "# MboI，为GATCGATC\n",
    "LIGATION_SITE = AAGCTAGCTT\n",
    "\n",
    "# Minimum size of restriction fragments to consider for the Hi-C processing.\n",
    "# Example: 100\n",
    "# 为Hi-C处理考虑的限制片段的【最小大小】\n",
    "MIN_FRAG_SIZE = 100\n",
    "\n",
    "# Maximum size of restriction fragments to consider for the Hi-C processing.\n",
    "# Example: 100000\n",
    "# # 为Hi-C处理考虑的限制片段的【最大大小】\n",
    "MAX_FRAG_SIZE = 100000\n",
    "\n",
    "# Minimum sequenced insert size. \n",
    "# Shorter 3C products are discarded. \n",
    "# Example: 100\n",
    "# 测得【最小】插入大小。\n",
    "# 较短的3C产物会被丢弃\n",
    "MIN_INSERT_SIZE = 100\n",
    "\n",
    "# Maximum sequenced insert size. \n",
    "# Larger 3C products are discarded. \n",
    "# Example: 600\n",
    "# 测得【最大】插入大小。\n",
    "# 较短的3C产物会被丢弃\n",
    "MAX_INSERT_SIZE = 600\n",
    "\n",
    "#######################################################################\n",
    "## Hi-C processing\n",
    "#######################################################################\n",
    "# Filter short range contact below the specified distance. \n",
    "# Mainly useful for DNase Hi-C. \n",
    "# Example: 1000\n",
    "MIN_CIS_DIST =\n",
    "\n",
    "# Create output files with all classes of 3C products. \n",
    "# Default: 0\n",
    "GET_ALL_INTERACTION_CLASSES = 1\n",
    "\n",
    "# Create a BAM file with all aligned reads flagged according to \n",
    "# their classifaction and mapping category. \n",
    "# 是否保留BAM文件，默认不保留\n",
    "# Default: 0\n",
    "GET_PROCESS_SAM = 1\n",
    "\n",
    "# Remove singleton reads. \n",
    "# Default: 1\n",
    "RM_SINGLETON = 1\n",
    "\n",
    "# Remove multi-mapped reads. \n",
    "# Default: 1\n",
    "RM_MULTI = 1\n",
    "\n",
    "# Remove duplicated reads’ pairs. \n",
    "# Default: 1\n",
    "RM_DUP = 1\n",
    "\n",
    "#######################################################################\n",
    "## Contact Maps\n",
    "#######################################################################\n",
    "# Resolution of contact maps to generate (space separated). \n",
    "# Default: 20000 40000 150000 500000 1000000\n",
    "BIN_SIZE = 20000 40000 150000 500000 1000000\n",
    "\n",
    "# Binning step size in ‘n’ coverage _i.e._ window step. \n",
    "# Default: 1\n",
    "# BIN_STEP\n",
    "\n",
    "# Output matrix format. \n",
    "# Must be complete, asis, upper or lower. \n",
    "# Default: upper\n",
    "MATRIX_FORMAT = upper\n",
    "\n",
    "#######################################################################\n",
    "## Normalization\n",
    "#######################################################################\n",
    "# Maximum number of iteration for ICE normalization. \n",
    "# Default: 100\n",
    "MAX_ITER = 100\n",
    "\n",
    "# Define which pourcentage of bins with low counts should be force to zero. \n",
    "# Default: 0.02. \n",
    "# Replace SPARSE_FILTERING\n",
    "FILTER_LOW_COUNT_PERC = 0.02\n",
    "\n",
    "# Define which pourcentage of bins with low counts should be discarded \n",
    "# before normalization. \n",
    "# Default: 0\n",
    "FILTER_HIGH_COUNT_PERC = 0\n",
    "\n",
    "# The relative increment in the results before declaring convergence. \n",
    "# Default: 0.1\n",
    "EPS = 0.1\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert ValidPairs to Juicer .hic\n",
    "~/0.apps/HiC-Pro_installed/HiC-Pro_3.1.0/bin/utils/hicpro2juicebox.sh -h\n",
    "提示需要Hi Doug,\n",
    "\n",
    "Please take a look at: \n",
    "\n",
    "https://github.com/theaidenlab/juicer/tree/master/AWS/scripts\n",
    "\n",
    "juicebox_tools.jar is located there.\n",
    "\n",
    "juicebox_tools.jar and juicebox_CLT.jar are different names for the same thing.\n",
    "If your script is using CLT, just rename that jar to the correct name.\n",
    "\n",
    "```shell\n",
    "~/0.apps/HiC-Pro_installed/HiC-Pro_3.1.0/bin/utils/hicpro2juicebox.sh -i dixon_2M.allValidPairs -g ~/0.apps/HiC-Pro_installed/HiC-Pro_3.1.0/annotation/chrom_hg38.sizes -j ~/0.apps/juicerbox/juicer_tools.jar -r ~/0.apps/HiC-Pro_installed/HiC-Pro_3.1.0/annotation/HindIII_resfrag_hg38.bed\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# call loop [juicer]\n",
    "https://github.com/aidenlab/juicer\n",
    "\n",
    "http://www.360doc.com/content/19/1224/14/68068867_881786243.shtml\n",
    "\n",
    "juicer采用ArrowHead算法对原始的交互矩阵进行转化，并预测TAD拓扑关联结构域，采用HiCUUPS算法识别染色质环chromatin loops。和其他Hi-C数据处理软件相比，juicer的功能更为齐全\n",
    "\n",
    "```shell\n",
    "Command Line Tools Usage\n",
    "Detailed documentation about the command line tools can be found on the wiki:\n",
    "\n",
    "Annotating features with Arrowhead, HiCCUPS, MotifFinder, APA, Eigenvector, and Pearsons\n",
    "Creating .hic with Pre\n",
    "Extracting data from .hic files with dump\n",
    "To launch the command line tools, use the shell script “juicer_tools” on Unix/MacOS or type\n",
    "\n",
    "java -jar juicer_tools.jar (command...) [flags...] <parameters...>`\n",
    "There are different flavors of juicer_tools that depend on the CUDA version. If you do not use GPUs, these versions are equivalent. Otherwise, juicer_tools.X.X.jar uses CUDA version X.X\n",
    "\n",
    "For HiCCUPS loop calling without the shell or bat script, you will need to call: java -Xms512m -Xmx2048m -Djava.library.path=path/to/natives/ -jar juicer_tools.jar hiccups [flags...] <parameters...> where path/to/natives is the path to the native libraries used for Jcuda By default, these are located in the lib/jcuda folder.\n",
    "\n",
    "In the command line tools, there are several analysis functions:\n",
    "\n",
    "apa for conducting aggregate peak analysis\n",
    "hiccups for annotating loops\n",
    "motifs for finding CTCF motifs\n",
    "arrowhead for annotating contact domains\n",
    "eigenvector for calculating the eigenvector (first PC) of the Pearson's\n",
    "pearsons for calculating the Pearson's\n",
    "The juicer_tools (Unix/MacOS) script can be used in place of the unwieldy java -Djava.library.path=path/to/natives/ -jar juicer_tools.jar\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# call TAD [HiC Explorer / TopDom]\n",
    "/lustre1/chengqiyi_pkuhpc/zhaohn/0.apps/miniconda3/bin/hicexplorer\n",
    "\n",
    "这个链接【重点参考！！！】\n",
    "https://blog.csdn.net/hzau_yang/article/details/100031590\n",
    "```shell\n",
    "hicBuildMatrix --samFiles mate_R1.bam mate_R2.bam \\\n",
    "                 --binSize 10000 \\\n",
    "                 --restrictionSequence GATC \\\n",
    "                 --threads 4\n",
    "                 --inputBufferSize 100000\n",
    "                 --outBam hic.bam \\\n",
    "                 -o hic_matrix.h5\n",
    "                 --QCfolder ./hicQC\n",
    "```\n",
    "\n",
    "该步骤最终生成一个bam文件，一个h5文件，以及一个hicQC文件夹报告，改报告写的比较详细，懂Hi-C的一般都能看懂，里面包含了比对率，数据有效利用率等信息；h5这个文件格式比较复杂，操作起来需要hicmatrix包的HiCMatrix函数；\n",
    "\n",
    "\n",
    "3.校正Hi-C交互矩阵\n",
    "默认使用KR标准化方法，也可以使用ICE，通过–correctionMethod参数控制；\n",
    "–chromosomes控制输出的染色体，通过空格隔开，如chr1 chr2 chr3;\n",
    "\n",
    "```shell\n",
    "# KR 校正\n",
    "hicCorrectMatrix correct -m hic_matrix.h5 --filterThreshold -1.5 5 -o hic_corrected.h5\n",
    "# 画热图\n",
    "hicPlotMatrix -m hic_corrected.h5 -o hic_plot.png --region 1:20000000-80000000 --log1p\n",
    "```\n",
    "该命令可以根据输出文件后缀画出不同的格式图，也可以通过–perChr参数控制每条染色体单独画；\n",
    "\n",
    "\n",
    "找TAD\n",
    "找TAD的软件有很多，hicexplorer的方法与Topdom有点相似，总的来说算比较简单粗暴的，实现\n",
    "hicFindTADs -m hic_corrected.h5 --outPrefix hic_corrected --numberOfProcessors 16\n",
    "\n",
    "# call compartments\n",
    "\n",
    "5.找compartment\n",
    "\n",
    "\n",
    "hicPCA -m hic_corrected.h5 --outFileName pca1.bw pca2.bw --format bigwig --pearsonMatrix pearson.h5\n",
    "\n",
    "通常可以根据基因密度来调整第一主成分的符号，获得最终的compartment；统计完每个bin的基因数量，得到bigwig文件，然后通过–extraTrack可以直接对PCA结果调整符号；也可以通过–pearsonMatrix和–obsexpMatrix生成计算compartment的中间处理中的矩阵，如pearson矩阵；\n",
    "\n",
    "画图\n",
    "\n",
    "hicPlotMatrix -m pearson.h5 --outFileName pca1.png --perChr --bigwig pca1.bw"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```shell\n",
    "###############################################################################################\n",
    "# 2021-04-06\n",
    "# Hi-C data analysis with Hi-C pro\n",
    "###############################################################################################\n",
    "# Hi-C in HEK293\n",
    "# Zuin J, Dixon JR, van der Reijden MI, Ye Z et al. Cohesin and CTCF differentially affect chromatin architecture and gene expression in human cells. Proc Natl Acad Sci U S A 2014 Jan 21;111(3):996-1001. PMID: 24335803\n",
    "# https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE44267\n",
    "\n",
    "GSM1081526\tHi-C, RAD21cv HEK293, TEV treated, replicate one\n",
    "GSM1081527\tHi-C, RAD21cv HEK293, TEV treated, replicate two\n",
    "GSM1081528\tHi-C, RAD21cv HEK293, HRV treated, replicate one\n",
    "GSM1081529\tHi-C, RAD21cv HEK293, HRV treated, replicate two\n",
    "GSM1081530\tHi-C, HEK293 siRNA Control, replicate one\n",
    "GSM1081531\tHi-C, HEK293 siRNA Control, replicate two\n",
    "GSM1081532\tHi-C, HEK293 siRNA CTCF, replicate one\n",
    "GSM1081533\tHi-C, HEK293 siRNA CTCF, replicate two\n",
    "GSM1081534\tmRNA-seq, RAD21cv HEK293, TEV treated, replicate one\n",
    "GSM1081535\tmRNA-seq, RAD21cv HEK293, TEV treated, replicate two\n",
    "GSM1081536\tmRNA-seq, RAD21cv HEK293, HRV treated, replicate one\n",
    "GSM1081537\tmRNA-seq, RAD21cv HEK293, HRV treated, replicate two\n",
    "GSM1081538\tmRNA-seq, HEK293 siRNA Control, replicate one\n",
    "GSM1081539\tmRNA-seq, HEK293 siRNA Control, replicate two\n",
    "GSM1081540\tmRNA-seq, HEK293 siRNA CTCF, replicate one\n",
    "GSM1081541\tmRNA-seq, HEK293 siRNA CTCF, replicate two\n",
    "GSM1081542\tSmc3 ChIP-seq, RAD21cv HEK293, TEV Treated\n",
    "GSM1081543\tInput ChIP-seq, RAD21cv HEK293, TEV Treated\n",
    "\n",
    "\n",
    "# fix filename \n",
    "SRR710071_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_1.fastq.gz\n",
    "SRR710071_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_2.fastq.gz\n",
    "SRR710072_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_1.fastq.gz\n",
    "SRR710072_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_2.fastq.gz\n",
    "SRR710073_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_1.fastq.gz\n",
    "SRR710073_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_2.fastq.gz\n",
    "SRR710074_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_1.fastq.gz\n",
    "SRR710074_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_2.fastq.gz\n",
    "SRR710075_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_1.fastq.gz\n",
    "SRR710075_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_2.fastq.gz\n",
    "SRR710076_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_1.fastq.gz\n",
    "SRR710076_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_2.fastq.gz\n",
    "SRR710077_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_1.fastq.gz\n",
    "SRR710077_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_2.fastq.gz\n",
    "\n",
    "\n",
    "cat SRR710071_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_1.fastq.gz \\\n",
    "SRR710072_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_1.fastq.gz \\\n",
    "SRR710073_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_1.fastq.gz \\\n",
    "SRR710074_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_1.fastq.gz \\\n",
    "> 293T-HiC-WT_rep1_R1.fq.gz & \n",
    "\n",
    "cat SRR710071_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_2.fastq.gz \\\n",
    "SRR710072_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_2.fastq.gz \\\n",
    "SRR710073_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_2.fastq.gz \\\n",
    "SRR710074_GSM1081530_Hi-C_HEK293_siRNA_Control_replicate_one_Homo_sapiens_OTHER_2.fastq.gz \\\n",
    "> 293T-HiC-WT_rep1_R2.fq.gz & \n",
    "\n",
    "cat SRR710075_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_1.fastq.gz \\\n",
    "SRR710076_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_1.fastq.gz \\\n",
    "SRR710077_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_1.fastq.gz \\\n",
    "> 293T-HiC-WT_rep2_R1.fq.gz & \n",
    "\n",
    "cat SRR710075_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_2.fastq.gz \\\n",
    "SRR710076_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_2.fastq.gz \\\n",
    "SRR710077_GSM1081531_Hi-C_HEK293_siRNA_Control_replicate_two_Homo_sapiens_OTHER_2.fastq.gz \\\n",
    "> 293T-HiC-WT_rep2_R2.fq.gz & \n",
    "\n",
    "\n",
    "mv 293T-HiC-WT_rep1_R1.fq.gz  293T-HiC-WT_rep1_R1.fastq.gz\n",
    "mv 293T-HiC-WT_rep1_R2.fq.gz  293T-HiC-WT_rep1_R2.fastq.gz\n",
    "mv 293T-HiC-WT_rep2_R1.fq.gz  293T-HiC-WT_rep2_R1.fastq.gz\n",
    "mv 293T-HiC-WT_rep2_R2.fq.gz  293T-HiC-WT_rep2_R2.fastq.gz\n",
    "\n",
    "mv 293T-HiC-WT_rep1_R1.fastq.gz 293T-HiC-WT-rep1_R1.fastq.gz\n",
    "mv 293T-HiC-WT_rep1_R2.fastq.gz 293T-HiC-WT-rep1_R2.fastq.gz\n",
    "mv 293T-HiC-WT_rep2_R1.fastq.gz 293T-HiC-WT-rep2_R1.fastq.gz\n",
    "mv 293T-HiC-WT_rep2_R2.fastq.gz 293T-HiC-WT-rep2_R2.fastq.gz\n",
    "\n",
    "\n",
    "\n",
    "# HiC pro prepare\n",
    "./digest_genome.py  -r A^AGCTT  -o /home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38_hindiii.bed ~/menghw_HD/reference/bowtie2_hg38/hg38_only_chromosome.fa &\n",
    "\n",
    "# make bash\n",
    "cd /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data\n",
    "\n",
    "HiC-Pro -i ./rawdata -o 01.hicpro_WT/ -c config-hicpro.txt\n",
    "\n",
    "# init\n",
    "SLURM_SUBMIT_DIR=/home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/01.hicpro_WT\n",
    "\n",
    "# run \n",
    "cat 293T-HiC-WT-rep1_R1.fastq.gz 293T-HiC-WT-rep2_R1.fastq.gz > 293T-HiC-WT-merge_R1.fastq.gz &\n",
    "cat 293T-HiC-WT-rep1_R2.fastq.gz 293T-HiC-WT-rep2_R2.fastq.gz > 293T-HiC-WT-merge_R2.fastq.gz &\n",
    "\n",
    "\n",
    "# HiC pro data to .hic format \n",
    "cd /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/01.hicpro_WT/hic_results/data/293T_WT\n",
    "\n",
    "hicpro2juicebox.sh \\\n",
    "-i 293T_WT.allValidPairs \\\n",
    "-g /home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38.only_chrom.sizes \\\n",
    "-j /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar \\\n",
    "-r /home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38_hindiii.bed > 293T_WT.allValidPairs.hicproTohic.log  2>&1 &\n",
    "\n",
    "\n",
    "hicpro2juicebox.sh \\\n",
    "-i test.allValidPairs \\\n",
    "-g /home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38.only_chrom.sizes \\\n",
    "-j /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar \\\n",
    "-r /home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38_hindiii.bed > test.allValidPairs.hicproTohic.log  2>&1  &\n",
    "\n",
    "\n",
    "java -Xmx64g -jar /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar pre \\\n",
    "--threads 20 -j 20 \\\n",
    "-f ./tmp/2404_resfrag.juicebox  \\\n",
    "-r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000 \\\n",
    "./tmp/2404_allValidPairs.pre_juicebox_sorted \\\n",
    "293T_WT.allValidPairs.NoFrag.hic \\\n",
    "/home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38.only_chrom.sizes > 293T_WT.allValidPairs.NoFrag.hicproTohic.log  2>&1 &\n",
    "\n",
    "\n",
    "# call loops \n",
    "cd /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/01.hicpro_WT/hic_results/data/293T_WT\n",
    "\n",
    "# java -Xmx32g -jar /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar hiccups \\\n",
    "# --threads 10 \\\n",
    "# -r 25000,10000,5000 \\\n",
    "# -k KR \\\n",
    "# -f 0.1,0.1,0.1 \\\n",
    "# -p 1,2,4 \\\n",
    "# -d 20000,20000,20000 \\\n",
    "# test.allValidPairs.hic \\\n",
    "\n",
    "\n",
    "java -Xmx64g -jar /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar hiccups \\\n",
    "--threads 20 \\\n",
    "-r 5000,10000,25000 \\\n",
    "-c chr1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X \\\n",
    "--ignore-sparsity \\\n",
    "293T_WT.allValidPairs.NoFrag.hic all_hiccups_loops.NoFrag.hic\n",
    "\n",
    "\n",
    "java -Xmx64g -jar /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar hiccups \\\n",
    "--threads 20 \\\n",
    "-r 5000,10000,25000 \\\n",
    "-c chr1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X \\\n",
    "--ignore-sparsity \\\n",
    "293T_WT.allValidPairs.hic all_hiccups_loops \n",
    "\n",
    "# extract data \n",
    "java -Xmx32g -jar /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar dump \\\n",
    "observed KR 293T_WT.allValidPairs.NoFrag.hic chr1 chr1 BP 100000 test.txt \n",
    "\t\n",
    "\n",
    "# java -Xmx32g -jar /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar hiccups \\\n",
    "# -r 5000 -c chr1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X --ignore-sparsity \\\n",
    "# 293T_WT.allValidPairs.NoFrag.hic all_hiccups_loops\n",
    "\n",
    "# # on Lab iMac\n",
    "# java -Xmx32g -jar /Users/meng/juicer_tools_1.22.01.jar hiccups \\\n",
    "# -r 5000,10000,25000 -c chr1,2,3,4,5,6,7,8,9,10 293T_WT.allValidPairs.hic all_hiccups_loops\n",
    "\n",
    "\n",
    "# cuda install \n",
    "https://developer.nvidia.com/cuda-10.2-download-archive?target_os=MacOSX&target_arch=x86_64&target_version=1013&target_type=dmglocal\n",
    "\n",
    "####################################################################################################\n",
    "# HiCExplore run \n",
    "####################################################################################################\n",
    "cd /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/01.hicpro_WT\n",
    "\n",
    "# hicExplore make restriction sites\n",
    "hicFindRestSite \\\n",
    "--fasta ~/menghw_HD/reference/bowtie2_hg38/hg38_only_chromosome.fa \\\n",
    "--searchPattern AAGCTT \\\n",
    "-o hicexp_result/hicexp_hindiii_hg38.bed &\n",
    "\n",
    "# build up hic matrix \n",
    "hicBuildMatrix \\\n",
    "--samFiles \\\n",
    "./bowtie_results/bwt2/293T_WT/293T-HiC-WT-merge_R1_hg38_only_chromosome.bwt2merged.bam \\\n",
    "./bowtie_results/bwt2/293T_WT/293T-HiC-WT-merge_R2_hg38_only_chromosome.bwt2merged.bam \\\n",
    "--outFileName ./hicexp_result/293T_WT.RawMatrix.h5 \\\n",
    "--binSize 5000 10000 25000 40000 50000 100000 250000 500000 1000000 \\\n",
    "--threads 16 \\\n",
    "--QCfolder ./hicexp_result/QC_res \\\n",
    "--restrictionCutFile ./hicexp_result/hicexp_hindiii_hg38.bed \\\n",
    "--restrictionSequence AAGCTT \\\n",
    "--danglingSequence AGCT \\\n",
    "--minMappingQuality 20 \\\n",
    "--chromosomeSizes /home/menghaowei/menghw_HD/reference/hg38.only_chrom.sizes  > ./hicexp_result/293T_WT.hicBuildMatrix.RawMatrix.log 2>&1 & \n",
    "\n",
    "# test\n",
    "hicBuildMatrix \\\n",
    "--samFiles \\\n",
    "./bowtie_results/bwt2/293T_WT/test_R1.sam \\\n",
    "./bowtie_results/bwt2/293T_WT/test_R2.sam \\\n",
    "--outFileName ./hicexp_result/test.RawMatrix.cool \\\n",
    "--binSize 50000 100000 250000 500000 1000000 \\\n",
    "--threads 4 \\\n",
    "--QCfolder ./hicexp_result/QC_res.test \\\n",
    "--restrictionCutFile ./hicexp_result/hicexp_hindiii_hg38.bed \\\n",
    "--restrictionSequence AAGCTT \\\n",
    "--danglingSequence AGCT \\\n",
    "--minMappingQuality 20 \\\n",
    "--chromosomeSizes /home/menghaowei/menghw_HD/reference/hg38.only_chrom.sizes  > ./hicexp_result/test.hicBuildMatrix.RawMatrix.log 2>&1 & \n",
    "\n",
    "#### correct matrix \n",
    "# hicCorrectMatrix diagnostic_plot -m test.RawMatrix.cool -o test.hic_corrected.png\n",
    "\n",
    "# hicCorrectMatrix correct -m test.RawMatrix.cool --filterThreshold -1.5 5 -o test.KRCorrectMatrix.h5 & \n",
    "\n",
    "hicCorrectMatrix correct -m 293T_WT.RawMatrix.h5 --filterThreshold -1.5 5 -o 293T_WT.KRCorrectMatrix.h5 > 293T_WT.hicCorrectMatrix.log 2>&1 & \n",
    "\n",
    "#### call TAD\n",
    "# hicFindTADs -m test.KRCorrectMatrix.h5 --outPrefix test.KRCorrectMatrix --numberOfProcessors 16 &  \n",
    "\n",
    "hicFindTADs -m 293T_WT.KRCorrectMatrix.h5 --outPrefix 293T_WT.KR --numberOfProcessors 16 --correctForMultipleTesting fdr > 293T_WT.hicFindTADs.log 2>&1 &  \n",
    "\n",
    "#### merge Hi-C matrix\n",
    "# merge hic bins with 25Kbp binsize \n",
    "hicMergeMatrixBins --matrix 293T_WT.RawMatrix.h5 -o 293T_WT.RawMatrix.25Kbp.h5 --numBins 5 & \n",
    "\n",
    "# merge hic bins with 50Kbp binsize \n",
    "hicMergeMatrixBins --matrix 293T_WT.RawMatrix.h5 -o 293T_WT.RawMatrix.50Kbp.h5 --numBins 10 & \n",
    "\n",
    "# correct 25Kbp\n",
    "hicCorrectMatrix correct -m 293T_WT.RawMatrix.25Kbp.h5 --correctionMethod KR \\\n",
    "--filterThreshold -1.5 5 -o 293T_WT.KR.CorrectMatrix.25Kbp.h5 > 293T_WT.hicCorrectMatrix.25Kbp.log 2>&1 & \n",
    "\n",
    "# correct 50Kbp\n",
    "hicCorrectMatrix correct -m 293T_WT.RawMatrix.50Kbp.h5 --correctionMethod KR \\\n",
    "--filterThreshold -1.5 5 -o 293T_WT.KR.CorrectMatrix.50Kbp.h5 > 293T_WT.KR.hicCorrectMatrix.50Kbp.log 2>&1 & \n",
    "\n",
    "# correct 50Kbp ICE\n",
    "hicCorrectMatrix correct -m 293T_WT.RawMatrix.50Kbp.h5 --correctionMethod ICE \\\n",
    "-o 293T_WT.ICE.CorrectMatrix.50Kbp.h5   \\\n",
    "--filterThreshold -1.5 5 --iterNum 500  \\\n",
    "> 293T_WT.ICE.hicCorrectMatrix.50Kbp.log 2>&1 & \n",
    "\n",
    "# find TAD \n",
    "srun -T 24 hicFindTADs -m 293T_WT.KR.CorrectMatrix.25Kbp.h5 --outPrefix 293T_WT.KR.25Kbp --numberOfProcessors 24 --correctForMultipleTesting fdr > 293T_WT.hicFindTADs.25Kbp.log 2>&1 &  \n",
    "\n",
    "srun -T 24 hicFindTADs -m 293T_WT.KR.CorrectMatrix.50Kbp.h5 --outPrefix 293T_WT.KR.50Kbp --numberOfProcessors 24 --correctForMultipleTesting fdr > 293T_WT.KR.hicFindTADs.50Kbp.log 2>&1 &  \n",
    "\n",
    "srun -T 24 hicFindTADs -m 293T_WT.ICE.CorrectMatrix.50Kbp.h5 --outPrefix 293T_WT.ICE.50Kbp --numberOfProcessors 24 --correctForMultipleTesting fdr > 293T_WT.hicFindTADs.50Kbp.log 2>&1 &  \n",
    "\n",
    "\n",
    "# plot TAD\n",
    "hicPlotTADs --tracks plot_TAD.ini -o out_image/293T_WT.KR.BinSize5Kbp.chr12_96M_126M.pdf --region chr12:96000000-126000000\n",
    "\n",
    "hicPlotTADs --tracks plot_TAD.DdCBEOnly.ini -o out_image/293T_WT.KR.BinSize5Kbp.chr12_96M_126M.DdCBEOnly.pdf --region chr12:96000000-126000000 &\n",
    "\n",
    "hicPlotTADs --tracks plot_TAD.DdCBEOnly.25Kbp.ini -o out_image/293T_WT.KR.BinSize25Kbp.chr12_96M_126M.DdCBEOnly.pdf --region chr12:96000000-126000000 &\n",
    "\n",
    "hicPlotTADs --tracks plot_TAD.DdCBEOnly.25Kbp.ini -o out_image/293T_WT.KR.BinSize25Kbp.chr12_96M_116M.DdCBEOnly.pdf --region chr12:96000000-116000000 &\n",
    "\n",
    "hicPlotTADs --tracks plot_TAD.Reds.DdCBEOnly.25Kbp.ini -o out_image/293T_WT.KR.BinSize25Kbp.chr12_96M_126M.DdCBEOnly.Reds.pdf --region chr12:96000000-126000000 &\n",
    "\n",
    "hicPlotTADs --tracks plot_TAD.DdCBEOnly.25Kbp.ini -o out_image/293T_WT.KR.BinSize25Kbp.chr12_10M_130M.DdCBEOnly.pdf --region chr12:10000000-130000000 &\n",
    "\n",
    "\n",
    "# plot TAD score \n",
    "sort -k1,1 -k2,2n 293T_WT.KR_score.bedgraph > 293T_WT.KR_score.sort.bedgraph &\n",
    "\n",
    "sort -k1,1 -k2,2n 293T_WT.KR.25Kbp_score.bedgraph > 293T_WT.KR.25Kbp_score.sort.bedgraph &\n",
    "\n",
    "# bedgraph to bigwig\n",
    "bedGraphToBigWig 293T_WT.KR_score.sort.bedgraph ~/menghw_HD/reference/hg38.only_chrom.sizes 293T_WT.KR_score.bigwig &\n",
    "\n",
    "bedGraphToBigWig 293T_WT.KR.25Kbp_score.sort.bedgraph ~/menghw_HD/reference/hg38.only_chrom.sizes 293T_WT.KR.25Kbp_score.bigwig &\n",
    "\n",
    "# region in /home/menghaowei/menghw_HD/DdCBE_project/08.DdCBE_merge_all/region_cor_analysis/peak_region\n",
    "\n",
    "# plot TAD\n",
    "cd /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/01.hicpro_WT/hicexp_result\n",
    "\n",
    "computeMatrix reference-point -S \\\n",
    "~/menghw_HD/DdCBE_project/10.hic_data/01.hicpro_WT/hicexp_result/01.TAD_info/293T_WT.KR.25Kbp_score.bigwig \\\n",
    "~/menghw_HD/DdCBE_project/10.hic_data/01.hicpro_WT/hicexp_result/01.TAD_info/293T_WT.KR_score.bigwig \\\n",
    "-R \\\n",
    "~/menghw_HD/DdCBE_project/08.DdCBE_merge_all/region_cor_analysis/peak_region/20210312-Motif-ND6-DepSite.merge.bed \\\n",
    "~/menghw_HD/DdCBE_project/08.DdCBE_merge_all/region_cor_analysis/peak_region/20210312-Motif-ND6-Indep.merge.bed \\\n",
    "~/menghw_HD/DdCBE_project/08.DdCBE_merge_all/region_cor_analysis/peak_region/20210319-hg38_random.50bp.sort.bed \\\n",
    "--referencePoint center \\\n",
    "--beforeRegionStartLength 2000000 \\\n",
    "--afterRegionStartLength 2000000 \\\n",
    "--skipZeros \\\n",
    "--binSize 25000 \\\n",
    "-o deeptools_result/20210421-TAD_score.profile.mat.gz \\\n",
    "--samplesLabel  TAD.25Kbp  TAD.5Kbp \\\n",
    "--numberOfProcessors 10 & \n",
    "\n",
    "# plot\n",
    "plotHeatmap -m deeptools_result/20210421-TAD_score.profile.mat.gz \\\n",
    "--colorMap Purples \\\n",
    "-out deeptools_result/20210421-TAD_score.profile.heatmap.pdf &\n",
    "\n",
    "# count site in TAD boundary\n",
    "Dep 3 / 72\n",
    "Indep 51 / 417 \n",
    "Random 16 / 500\n",
    "\n",
    "# count site in TAD boundary +- 1 pixel\n",
    "Dep 8 / 72\n",
    "Indep 134 / 417 \n",
    "Random 58 / 500\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "# 2021-10-10\n",
    "# Hi-C test data for qulity control \n",
    "####################################################################################################\n",
    "\n",
    "# HiC pro prepare DpnII\n",
    "digest_genome.py  -r ^GATC  -o /home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38_dpnii.bed ~/menghw_HD/reference/bowtie2_hg38/hg38_only_chromosome.fa &\n",
    "\n",
    "ligation_site: GATCGATC\n",
    "\n",
    "# rename hic data\n",
    "mv Hic-GFP-HiC-rep1_FKDL210259733-1a-1_1.fq.gz  293T_GFP_rep1/293T-HiC-batch_test-GFP_rep1_R1.fastq.gz\n",
    "mv Hic-GFP-HiC-rep1_FKDL210259733-1a-1_2.fq.gz  293T_GFP_rep1/293T-HiC-batch_test-GFP_rep1_R2.fastq.gz\n",
    "\n",
    "mv Hic-GFP-HiC-rep2_FKDL210259733-1a-2_1.fq.gz  293T_GFP_rep2/293T-HiC-batch_test-GFP_rep2_R1.fastq.gz\n",
    "mv Hic-GFP-HiC-rep2_FKDL210259733-1a-2_2.fq.gz  293T_GFP_rep2/293T-HiC-batch_test-GFP_rep2_R2.fastq.gz\n",
    "\n",
    "mv Hic-ND6-WT-HiC-rep1_FKDL210259733-1a-3_1.fq.gz  293T_DdCBE_ND6_rep1/293T-HiC-batch_test-DdCBE_ND6_rep1_R1.fastq.gz\n",
    "mv Hic-ND6-WT-HiC-rep1_FKDL210259733-1a-3_2.fq.gz  293T_DdCBE_ND6_rep1/293T-HiC-batch_test-DdCBE_ND6_rep1_R2.fastq.gz\n",
    "\n",
    "mv Hic-ND6-WT-HiC-rep2_FKDL210259733-1a-4_1.fq.gz  293T_DdCBE_ND6_rep2/293T-HiC-batch_test-DdCBE_ND6_rep2_R1.fastq.gz\n",
    "mv Hic-ND6-WT-HiC-rep2_FKDL210259733-1a-4_2.fq.gz  293T_DdCBE_ND6_rep2/293T-HiC-batch_test-DdCBE_ND6_rep2_R2.fastq.gz\n",
    "\n",
    "# fix config txt file \n",
    "config-hicpro_hic_test_DpnII.txt\n",
    "\n",
    "# make bash\n",
    "cd /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data\n",
    "\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/Data_Backup/DdCBE_project/08.hic_test_data.20211010/rename_fq \\\n",
    "-o 02.hic_test_DdCBE \\\n",
    "-c config-hicpro_hic_test_DpnII.txt &\n",
    "\n",
    "####################################################################################################\n",
    "# 2021-10-21\n",
    "# Hi-C MGI add data\n",
    "####################################################################################################\n",
    "dir = /home/menghaowei/menghw_HD/Data_Backup/DdCBE_project/09.hic_MGI_data.20211020\n",
    "\n",
    "61G ./V350027555/L03/V350027555_L03_read_1.fq.gz\n",
    "60G ./V350027555/L04/V350027555_L04_read_1.fq.gz\n",
    "60G ./V350027555/L01/V350027555_L01_read_1.fq.gz\n",
    "62G ./V350027555/L02/V350027555_L02_read_1.fq.gz\n",
    "63G ./V350027988/L03/V350027988_L03_read_1.fq.gz\n",
    "63G ./V350027988/L04/V350027988_L04_read_1.fq.gz\n",
    "62G ./V350027988/L01/V350027988_L01_read_1.fq.gz\n",
    "62G ./V350027988/L02/V350027988_L02_read_1.fq.gz\n",
    "\n",
    "# test \n",
    "srun -T 24 pigz -p 24 --test ./V350027555/L03/V350027555_L03_read_1.fq.gz &\n",
    "srun -T 24 pigz -p 24 --test ./V350027555/L01/V350027555_L01_read_1.fq.gz & \n",
    "srun -T 24 pigz -p 24 --test ./V350027988/L03/V350027988_L03_read_1.fq.gz &\n",
    "srun -T 24 pigz -p 24 --test ./V350027988/L01/V350027988_L01_read_1.fq.gz &\n",
    "srun -T 24 pigz -p 24 --test ./V350027555/L01/V350027555_L01_read_2.fq.gz &\n",
    "srun -T 24 pigz -p 24 --test ./V350027988/L01/V350027988_L01_read_2.fq.gz &\n",
    "\n",
    "\n",
    "# run Hi-C pro G1\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/Data_Backup/DdCBE_project/09.hic_MGI_data.20211020/rename_fq \\\n",
    "-o 03_1.hic_DdCBE_G1 \\\n",
    "-c config-hicpro_hic_MGI_G1.txt &\n",
    "\n",
    "\n",
    "# run Hi-C pro G2\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/Data_Backup/DdCBE_project/09.hic_MGI_data.20211020/rename_fq \\\n",
    "-o 03_1.hic_DdCBE_G2 \\\n",
    "-c config-hicpro_hic_MGI_G2.txt &\n",
    "\n",
    "\n",
    "# run Hi-C pro W1\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/Data_Backup/DdCBE_project/09.hic_MGI_data.20211020/rename_fq \\\n",
    "-o 03_1.hic_DdCBE_W1 \\\n",
    "-c config-hicpro_hic_MGI_W1.txt &\n",
    "\n",
    "\n",
    "# run Hi-C pro W2\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/Data_Backup/DdCBE_project/09.hic_MGI_data.20211020/rename_fq \\\n",
    "-o 03_1.hic_DdCBE_W2 \\\n",
    "-c config-hicpro_hic_MGI_W2.txt &\n",
    "\n",
    "#  \n",
    "03_1.hic_DdCBE_G1/bowtie_results/bwt2\n",
    "293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.bam  293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.bam\n",
    "samtools view -h test.bam | sed 's?/[1-2]??'\n",
    "\n",
    "# make test data for step run [HiC-Pro]\n",
    "samtools view  -h 293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.bam | head -n 1000000 | samtools view -hb -o /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/test/bowtie_results/bwt2/G1_fq/293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.bam\n",
    "samtools view  -h 293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.bam | head -n 1000000 | samtools view -hb -o /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/test/bowtie_results/bwt2/G1_fq/293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.bam\n",
    "\n",
    "\n",
    "MY_INSTALL_PATH/bin/HiC-Pro -i FULL_PATH_TO_DATA_FOLDER -o FULL_PATH_TO_OUTPUTS -c MY_LOCAL_CONFIG_FILE\n",
    "MY_INSTALL_PATH/bin/HiC-Pro -i FULL_PATH_TO_DATA_FOLDER -o FULL_PATH_TO_OUTPUTS -c MY_LOCAL_CONFIG_FILE -p\n",
    "\n",
    "/home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/test/bowtie_results/bwt2/G1_fq\n",
    "\n",
    "\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/test/bowtie_results/bwt2/G1_fq \\\n",
    "-o test \\\n",
    "-c test_G1.txt -s proc_hic -s quality_checks -s build_contact_maps -s ice_norm &\n",
    "\n",
    "\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/test/bowtie_results/bwt2/G1_fq/ \\\n",
    "-o test \\\n",
    "-c test_G1.txt -s proc_hic &\n",
    "\n",
    "\n",
    "/gpfs/user/menghaowei/anaconda2/envs/hicpro/bin/python \\\n",
    "/home/menghaowei/menghw_HD/software_package/HiC-Pro_3.0.0/scripts/mergeSAM.py \\\n",
    "-q 10 -t -v \\\n",
    "-f 293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.bam \\\n",
    "-r 293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.bam \\\n",
    "-o test.bwt2pairs.bam\n",
    "\n",
    "\n",
    "/gpfs/user/menghaowei/anaconda2/envs/hicpro/bin/python /home/menghaowei/menghw_HD/software_package/HiC-Pro_3.0.0/scripts/mergeSAM.py \\\n",
    "-q 10 -t -v \\\n",
    "-f 293T-HiC-batch_test-DdCBE_ND6_rep1_R1_hg38_only_chromosome.bwt2merged.bam \\\n",
    "-r 293T-HiC-batch_test-DdCBE_ND6_rep1_R2_hg38_only_chromosome.bwt2merged.bam \\\n",
    "-o test.bam\n",
    "\n",
    "# header \n",
    "less -S L01/V350027555_L01_read_1.fq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L1C001R0030000000/1\n",
    "@V350027555L1C001R0030000001/1\n",
    "@V350027555L1C001R0030000002/1\n",
    "@V350027555L1C001R0030000003/1\n",
    "@V350027555L1C001R0030000004/1\n",
    "@V350027555L1C001R0030000005/1\n",
    "@V350027555L1C001R0030000008/1\n",
    "@V350027555L1C001R0030000009/1\n",
    "@V350027555L1C001R0030000010/1\n",
    "@V350027555L1C001R0030000011/1\n",
    "\n",
    "less -S L01/V350027555_L01_read_2.fq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L1C001R0030000000/2\n",
    "@V350027555L1C001R0030000001/2\n",
    "@V350027555L1C001R0030000002/2\n",
    "@V350027555L1C001R0030000003/2\n",
    "@V350027555L1C001R0030000004/2\n",
    "@V350027555L1C001R0030000005/2\n",
    "@V350027555L1C001R0030000008/2\n",
    "@V350027555L1C001R0030000009/2\n",
    "@V350027555L1C001R0030000010/2\n",
    "@V350027555L1C001R0030000011/2\n",
    "\n",
    "less -S L02/V350027555_L02_read_1.fq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L2C001R0020000000/1\n",
    "@V350027555L2C001R0020000001/1\n",
    "@V350027555L2C001R0020000002/1\n",
    "@V350027555L2C001R0020000003/1\n",
    "@V350027555L2C001R0020000004/1\n",
    "@V350027555L2C001R0020000005/1\n",
    "@V350027555L2C001R0020000006/1\n",
    "@V350027555L2C001R0020000007/1\n",
    "@V350027555L2C001R0020000008/1\n",
    "@V350027555L2C001R0020000009/1\n",
    "\n",
    "\n",
    "less -S L02/V350027555_L02_read_2.fq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L2C001R0020000000/2\n",
    "@V350027555L2C001R0020000001/2\n",
    "@V350027555L2C001R0020000002/2\n",
    "@V350027555L2C001R0020000003/2\n",
    "@V350027555L2C001R0020000004/2\n",
    "@V350027555L2C001R0020000005/2\n",
    "@V350027555L2C001R0020000006/2\n",
    "@V350027555L2C001R0020000007/2\n",
    "@V350027555L2C001R0020000008/2\n",
    "@V350027555L2C001R0020000009/2\n",
    "\n",
    "\n",
    "less -S V350027555_L03_read_1.fq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L3C001R0030000001/1\n",
    "@V350027555L3C001R0030000002/1\n",
    "@V350027555L3C001R0030000003/1\n",
    "@V350027555L3C001R0030000004/1\n",
    "@V350027555L3C001R0030000005/1\n",
    "@V350027555L3C001R0030000006/1\n",
    "@V350027555L3C001R0030000007/1\n",
    "@V350027555L3C001R0030000008/1\n",
    "@V350027555L3C001R0030000009/1\n",
    "@V350027555L3C001R0030000010/1\n",
    "\n",
    "less -S V350027555_L03_read_2.fq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L3C001R0030000001/2\n",
    "@V350027555L3C001R0030000002/2\n",
    "@V350027555L3C001R0030000003/2\n",
    "@V350027555L3C001R0030000004/2\n",
    "@V350027555L3C001R0030000005/2\n",
    "@V350027555L3C001R0030000006/2\n",
    "@V350027555L3C001R0030000007/2\n",
    "@V350027555L3C001R0030000008/2\n",
    "@V350027555L3C001R0030000009/2\n",
    "@V350027555L3C001R0030000010/2\n",
    "\n",
    "\n",
    "less -S L04/V350027555_L04_read_1.fq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L4C001R0010000000/1\n",
    "@V350027555L4C001R0010000003/1\n",
    "@V350027555L4C001R0010000004/1\n",
    "@V350027555L4C001R0010000005/1\n",
    "@V350027555L4C001R0010000007/1\n",
    "@V350027555L4C001R0010000008/1\n",
    "@V350027555L4C001R0010000009/1\n",
    "@V350027555L4C001R0010000010/1\n",
    "@V350027555L4C001R0010000011/1\n",
    "@V350027555L4C001R0010000012/1\n",
    "\n",
    "\n",
    "less -S L04/V350027555_L04_read_2.fq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L4C001R0010000000/2\n",
    "@V350027555L4C001R0010000003/2\n",
    "@V350027555L4C001R0010000004/2\n",
    "@V350027555L4C001R0010000005/2\n",
    "@V350027555L4C001R0010000007/2\n",
    "@V350027555L4C001R0010000008/2\n",
    "@V350027555L4C001R0010000009/2\n",
    "@V350027555L4C001R0010000010/2\n",
    "@V350027555L4C001R0010000011/2\n",
    "@V350027555L4C001R0010000012/2\n",
    "\n",
    "\n",
    "less -S 293T-MGI-HiC-GFP_rep1_R1.fastq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L1C001R0030000000/1\n",
    "@V350027555L1C001R0030000001/1\n",
    "@V350027555L1C001R0030000002/1\n",
    "@V350027555L1C001R0030000003/1\n",
    "@V350027555L1C001R0030000004/1\n",
    "@V350027555L1C001R0030000005/1\n",
    "@V350027555L1C001R0030000008/1\n",
    "@V350027555L1C001R0030000009/1\n",
    "@V350027555L1C001R0030000010/1\n",
    "@V350027555L1C001R0030000011/1\n",
    "\n",
    "less -S 293T-MGI-HiC-GFP_rep1_R2.fastq.gz | sed '/^@V35/p' -n  | head -n 10\n",
    "@V350027555L1C001R0030000000/2\n",
    "@V350027555L1C001R0030000001/2\n",
    "@V350027555L1C001R0030000002/2\n",
    "@V350027555L1C001R0030000003/2\n",
    "@V350027555L1C001R0030000004/2\n",
    "@V350027555L1C001R0030000005/2\n",
    "@V350027555L1C001R0030000008/2\n",
    "@V350027555L1C001R0030000009/2\n",
    "@V350027555L1C001R0030000010/2\n",
    "@V350027555L1C001R0030000011/2\n",
    "\n",
    "\n",
    "samtools view -h 293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.bam | sed '/^V35/p' -n  | head -n 10 | cut -f 1\n",
    "V350027555L1C001R0010000000\n",
    "V350027555L1C001R0010000001\n",
    "V350027555L1C001R0010000002\n",
    "V350027555L1C001R0010000003\n",
    "V350027555L1C001R0010000004\n",
    "V350027555L1C001R0010000007\n",
    "V350027555L1C001R0010000008\n",
    "V350027555L1C001R0010000009\n",
    "V350027555L1C001R0010000010\n",
    "V350027555L1C001R0010000011\n",
    "\n",
    "samtools view -h 293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.bam | sed '/^V35/p' -n  | head -n 10 | cut -f 1\n",
    "V350027555L1C001R0030000000\n",
    "V350027555L1C001R0030000001\n",
    "V350027555L1C001R0030000002\n",
    "V350027555L1C001R0030000003\n",
    "V350027555L1C001R0030000004\n",
    "V350027555L1C001R0030000005\n",
    "V350027555L1C001R0030000008\n",
    "V350027555L1C001R0030000009\n",
    "V350027555L1C001R0030000010\n",
    "V350027555L1C001R0030000011\n",
    "\n",
    "\n",
    "samtools view -h 293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.RawName.bam | sed '/^V35/p' -n  | head -n 10 | cut -f 1\n",
    "V350027555L1C001R0010000000/1\n",
    "V350027555L1C001R0010000001/1\n",
    "V350027555L1C001R0010000002/1\n",
    "V350027555L1C001R0010000003/1\n",
    "V350027555L1C001R0010000004/1\n",
    "V350027555L1C001R0010000007/1\n",
    "V350027555L1C001R0010000008/1\n",
    "V350027555L1C001R0010000009/1\n",
    "V350027555L1C001R0010000010/1\n",
    "V350027555L1C001R0010000011/1\n",
    "\n",
    "samtools view -h 293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.RawName.bam | sed '/^V35/p' -n  | head -n 10 | cut -f 1\n",
    "V350027555L1C001R0030000000/2\n",
    "V350027555L1C001R0030000001/2\n",
    "V350027555L1C001R0030000002/2\n",
    "V350027555L1C001R0030000003/2\n",
    "V350027555L1C001R0030000004/2\n",
    "V350027555L1C001R0030000005/2\n",
    "V350027555L1C001R0030000008/2\n",
    "V350027555L1C001R0030000009/2\n",
    "V350027555L1C001R0030000010/2\n",
    "V350027555L1C001R0030000011/2\n",
    "\n",
    "9400043\tV350027555L1C001R0010000000/2\t0\tchr2\t203423038\t42\t84M\t*\t0\t0\tAGTGGCTTGCTAGTTTAGGGTACTTAATTGTTAGGTGTGGTTGCTAATACCCTACAATATTTGGTGGTCAGTCTTTTTAGTTCA\tFFFFFFEFFFFFFBFFEFFFFFFGFFFFFFEFFFFFFFFD>FFDFEFEFFEFFFFFFEGDFFFDFFFCFFFAAFEBFFBFEFFC\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:84\tYT:Z:UU\tRG:Z:BML\n",
    "\n",
    "# test \n",
    "293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.RawName.bam\n",
    "samtools view -h 293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.RawName.bam | head -n 15000000 | samtools view -hb > test_R2.RawName.bam &\n",
    "samtools view -h 293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.bam | head -n 15000000 | samtools view -hb > test_R2.FixName.bam &\n",
    "\n",
    "# sort by name \n",
    "samtools sort -@ 4 -T test_R2.RawName -m 10G -n -O BAM -o test_R2.RawName.SortName.bam test_R2.RawName.bam & \n",
    "samtools sort -@ 4 -T test_R2.FixName -m 10G -n -O BAM -o test_R2.FixName.SortName.bam test_R2.FixName.bam & \n",
    "\n",
    "# cmp reads name \n",
    "samtools view -h test_R2.RawName.bam | sed '/^V35/p' -n  | head -n 10 | cut -f 1\n",
    "V350027555L1C001R0030000000/2\n",
    "V350027555L1C001R0030000001/2\n",
    "V350027555L1C001R0030000002/2\n",
    "V350027555L1C001R0030000003/2\n",
    "V350027555L1C001R0030000004/2\n",
    "V350027555L1C001R0030000005/2\n",
    "V350027555L1C001R0030000008/2\n",
    "V350027555L1C001R0030000009/2\n",
    "V350027555L1C001R0030000010/2\n",
    "V350027555L1C001R0030000011/2\n",
    "\n",
    "samtools view -h test_R2.RawName.SortName.bam | sed '/^V35/p' -n  | head -n 10 | cut -f 1\n",
    "V350027555L1C001R0010000000/2\n",
    "V350027555L1C001R0010000001/2\n",
    "V350027555L1C001R0010000002/2\n",
    "V350027555L1C001R0010000003/2\n",
    "V350027555L1C001R0010000004/2\n",
    "V350027555L1C001R0010000007/2\n",
    "V350027555L1C001R0010000008/2\n",
    "V350027555L1C001R0010000009/2\n",
    "V350027555L1C001R0010000010/2\n",
    "V350027555L1C001R0010000011/2\n",
    "\n",
    "\n",
    "V350027555L1C001R0010000000\n",
    "V350027555L1C001R0010000001\n",
    "V350027555L1C001R0010000002\n",
    "V350027555L1C001R0010000003\n",
    "V350027555L1C001R0010000004\n",
    "V350027555L1C001R0010000007\n",
    "V350027555L1C001R0010000008\n",
    "V350027555L1C001R0010000009\n",
    "V350027555L1C001R0010000010\n",
    "V350027555L1C001R0010000011\n",
    "(py3env) menghaowei@cn07:G1_fq$samtools view 293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.bam | sed '/^V35/p' -n  | head -n 100 | cut -f 1\n",
    "V350027555L1C001R0010000000  V350027555L1C001R0010000000\n",
    "V350027555L1C001R0010000001  V350027555L1C001R0010000001\n",
    "V350027555L1C001R0010000002  V350027555L1C001R0010000002\n",
    "V350027555L1C001R0010000003  V350027555L1C001R0010000003\n",
    "V350027555L1C001R0010000004  V350027555L1C001R0010000004\n",
    "\n",
    "# test merge \n",
    "/gpfs/user/menghaowei/anaconda2/envs/hicpro/bin/python /home/menghaowei/menghw_HD/software_package/HiC-Pro_3.0.0/scripts/mergeSAM.py -q 10 -t -v -f bowtie_results/bwt2/W1_fq/293T-MGI-HiC-ND6_rep1_R1_hg38_only_chromosome.bwt2merged.bam -r bowtie_results/bwt2/W1_fq/293T-MGI-HiC-ND6_rep1_R2_hg38_only_chromosome.bwt2merged.bam -o bowtie_results/bwt2/W1_fq/293T-MGI-HiC-_hg38_only_chromosome.bwt2pairs.bam\n",
    "\n",
    "\n",
    "# test step \n",
    "HiC-Pro -i 03_1.hic_DdCBE_G1/bowtie_results/bwt2/G1_fq/293T-MGI-HiC-GFP_rep1_hg38_only_chromosome.bwt2pairs.bam \\\n",
    "-o 03_1.hic_DdCBE_G1 \\\n",
    "-c config-hicpro_hic_MGI_G1.txt -s proc_hic -s quality_checks -s build_contact_maps -s ice_norm &\n",
    "\n",
    "####################################################################################################\n",
    "# 2021-10-26\n",
    "# Hi-C MGI \n",
    "####################################################################################################\n",
    "/home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03.hic_DdCBE_MGI\n",
    "\n",
    ".\n",
    "├── [4.0K]  bam.hicpro\n",
    "├── [4.0K]  bam.hicpro.G1\n",
    "│   └── [4.0K]  GFP_rep1\n",
    "│       ├── [105G]  293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.SortName.bam\n",
    "│       └── [103G]  293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.SortName.bam\n",
    "├── [4.0K]  bam.hicpro.G2\n",
    "│   └── [4.0K]  GFP_rep2\n",
    "│       ├── [102G]  293T-MGI-HiC-GFP_rep2_R1_hg38_only_chromosome.bwt2merged.SortName.bam\n",
    "│       └── [100G]  293T-MGI-HiC-GFP_rep2_R2_hg38_only_chromosome.bwt2merged.SortName.bam\n",
    "├── [4.0K]  bam.hicpro.W1\n",
    "│   └── [4.0K]  ND6_rep1\n",
    "│       ├── [106G]  293T-MGI-HiC-ND6_rep1_R1_hg38_only_chromosome.bwt2merged.SortName.bam\n",
    "│       └── [108G]  293T-MGI-HiC-ND6_rep1_R2_hg38_only_chromosome.bwt2merged.SortName.bam\n",
    "└── [4.0K]  bam.hicpro.W2\n",
    "    └── [4.0K]  ND6_rep2\n",
    "        ├── [109G]  293T-MGI-HiC-ND6_rep2_R1_hg38_only_chromosome.bwt2merged.SortName.bam\n",
    "        └── [108G]  293T-MGI-HiC-ND6_rep2_R2_hg38_only_chromosome.bwt2merged.SortName.bam\n",
    "\n",
    "# run hicpro\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03.hic_DdCBE_MGI/bam.hicpro.G1 \\\n",
    "-o 03_1.hic_DdCBE_G1 \\\n",
    "-c hicpro_config/config-hicpro_hic_MGI_BAM_start.20211026.V2.txt \\\n",
    "-s proc_hic -s quality_checks -s build_contact_maps -s ice_norm &\n",
    "\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03.hic_DdCBE_MGI/bam.hicpro.G2 \\\n",
    "-o 03_1.hic_DdCBE_G2 \\\n",
    "-c hicpro_config/config-hicpro_hic_MGI_BAM_start.20211026.V2.txt \\\n",
    "-s proc_hic -s quality_checks -s build_contact_maps -s ice_norm &\n",
    "\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03.hic_DdCBE_MGI/bam.hicpro.W1 \\\n",
    "-o 03_1.hic_DdCBE_W1 \\\n",
    "-c hicpro_config/config-hicpro_hic_MGI_BAM_start.20211026.V2.txt \\\n",
    "-s proc_hic -s quality_checks -s build_contact_maps -s ice_norm &\n",
    "\n",
    "HiC-Pro -i /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03.hic_DdCBE_MGI/bam.hicpro.W2 \\\n",
    "-o 03_1.hic_DdCBE_W2 \\\n",
    "-c hicpro_config/config-hicpro_hic_MGI_BAM_start.20211026.V2.txt \\\n",
    "-s proc_hic -s quality_checks -s build_contact_maps -s ice_norm &\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "# 2021-10-26\n",
    "# Hi-C MGI data with HiC-Pro\n",
    "# 整理hicpro结果 并运行juicer\n",
    "####################################################################################################\n",
    "cd /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03.hic_DdCBE_MGI/hicpro_result/valid_pairs\n",
    "\n",
    "##### mkdir \n",
    "mkdir GFP_rep1\n",
    "mkdir GFP_rep2\n",
    "mkdir ND6_rep1\n",
    "mkdir ND6_rep2\n",
    "\n",
    "##### R1 and R2 mapping BAM\n",
    "# GFP_rep1\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_G1/hic_results/data/G1_fq/* ./\n",
    "\n",
    "# GFP_rep2\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_G2/hic_results/data/G2_fq/*  ./\n",
    "\n",
    "# WT_rep1\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_W1/hic_results/data/W1_fq/*  ./\n",
    "\n",
    "# WT_rep2\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_W2/hic_results/data/W2_fq/*  ./\n",
    "\n",
    "\n",
    "##### logs \n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_G1/logs/G1_fq/*  ./GFP_rep1/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_G2/logs/G2_fq/*  ./GFP_rep2/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_W1/logs/W1_fq/*  ./ND6_rep1/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_W2/logs/W2_fq/*  ./ND6_rep2/\n",
    "\n",
    "##### QC pic\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_G1/hic_results/pic/G1_fq/*  ./GFP_rep1/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_G2/hic_results/pic/G2_fq/*  ./GFP_rep2/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_W1/hic_results/pic/W1_fq/*  ./ND6_rep1/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_W2/hic_results/pic/W2_fq/*  ./ND6_rep2/\n",
    "\n",
    "\n",
    "##### merged BAM file \n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_G1/bowtie_results/bwt2/G1_fq/*  ./GFP_rep1/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_G2/bowtie_results/bwt2/G2_fq/*  ./GFP_rep2/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_W1/bowtie_results/bwt2/W1_fq/*  ./ND6_rep1/\n",
    "mv /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03_1.hic_DdCBE_W2/bowtie_results/bwt2/W2_fq/*  ./ND6_rep2/\n",
    "\n",
    "\n",
    "################################################################################\n",
    "# convert HiC-Pro format to juicer format\n",
    "################################################################################\n",
    "##### 原始脚本\n",
    "# 输入文件为 test.validPairs  则输出文件为test.validPairs.hic\n",
    "hicpro2juicebox.sh \\\n",
    "-i test.validPairs \\\n",
    "-g /home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38.only_chrom.sizes \\\n",
    "-j /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar \\\n",
    "-r /home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38_dpnii.bed > test.ValidPairs.hicproTojuicer.log  2>&1  &\n",
    "\n",
    "##### 问题\n",
    "1. 内含sort命令，可能需要大量内存；\n",
    "2. 最后一步调用juicer pre命令，内含1kb resolution的构成，可能会造成巨大的内存开销\n",
    "\n",
    "##### step1 from HiC-Pro validPairs to jucier pairs\n",
    "# fix hicpro2juicebox.sh script\n",
    "# 通过修改，可以分步骤运行\n",
    "\n",
    "\n",
    "##### step2 from juicer pairs to .hic file\n",
    "java -Xmx64g -jar /home/menghaowei/menghw_HD/software_package/juicer/juicer_tools_1.22.01.jar pre \\\n",
    "--threads 20 -j 20 \\\n",
    "-f ./tmp/2404_resfrag.juicebox  \\\n",
    "-r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000 \\\n",
    "./tmp/2404_allValidPairs.pre_juicebox_sorted \\\n",
    "293T_WT.allValidPairs.NoFrag.hic \\\n",
    "/home/menghaowei/menghw_HD/reference/HiCPro_ref/hg38.only_chrom.sizes > 293T_WT.allValidPairs.NoFrag.hicproTohic.log  2>&1 &\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "# 2021-10-29\n",
    "# Hi-C MGI data with HicExplorer\n",
    "####################################################################################################\n",
    "hicFindRestSite\n",
    "\n",
    "# hicExplore make restriction sites\n",
    "hicFindRestSite \\\n",
    "--fasta ~/menghw_HD/reference/bowtie2_hg38/hg38_only_chromosome.fa \\\n",
    "--searchPattern GATC \\\n",
    "-o /home/menghaowei/menghw_HD/reference/HicExplorer_ref/hicexp_dpnii_hg38.bed &\n",
    "\n",
    "\n",
    "# make matrix test code \n",
    "cd /home/menghaowei/menghw_HD/DdCBE_project/10.hic_data/03.hic_DdCBE_MGI/bam.hicpro\n",
    "\n",
    "# need about 150GB memory \n",
    "hicBuildMatrix \\\n",
    "--samFiles \\\n",
    "293T-MGI-HiC-GFP_rep1_R1_hg38_only_chromosome.bwt2merged.SortName.bam \\\n",
    "293T-MGI-HiC-GFP_rep1_R2_hg38_only_chromosome.bwt2merged.SortName.bam \\\n",
    "--outFileName test.RawMatrix.h5 \\\n",
    "--binSize 20000 40000 100000 200000 500000 1000000 \\\n",
    "--threads 10 \\\n",
    "--QCfolder QC_res \\\n",
    "--restrictionCutFile ~/menghw_HD/reference/HicExplorer_ref/hicexp_dpnii_hg38.bed \\\n",
    "--restrictionSequence GATC \\\n",
    "--danglingSequence GATC \\\n",
    "--minMappingQuality 20 \\\n",
    "--chromosomeSizes /home/menghaowei/menghw_HD/reference/hg38.only_chrom.sizes\n",
    "\n",
    "# matrix correction with KR methods\n",
    "# correct 20Kbp\n",
    "hicCorrectMatrix correct \\\n",
    "-m 293T-MGI-HiC-ND6_rep2_hg38.RawMatrix.h5 \\\n",
    "--correctionMethod KR \\\n",
    "-o 293T-MGI-HiC-ND6_rep2_hg38.KR.20Kbp.h5 > 293T-MGI-HiC-ND6_rep2_hg38.KR.20Kbp.log 2>&1 & \n",
    "\n",
    "# call TAD\n",
    "srun -T 24 hicFindTADs \\\n",
    "-m 293T_WT.KR.CorrectMatrix.25Kbp.h5 \\\n",
    "--outPrefix 293T_WT.KR.25Kbp \\\n",
    "--numberOfProcessors 24 \\\n",
    "--correctForMultipleTesting fdr > 293T_WT.hicFindTADs.25Kbp.log 2>&1 &  \n",
    "\n",
    "# plot \n",
    "hicPlotTADs --tracks 20211101-DdCBE_plot_TAD.ini -o out_image/test.pdf --region chrX:60000000-80000000 &\n",
    "hicPlotTADs --tracks 20211101-DdCBE_plot_TAD.ini -o out_image/test_2.pdf --region chrX:60000000-80000000 &\n",
    "hicPlotTADs --tracks 20211101-DdCBE_plot_TAD.ini -o out_image/test_3.pdf --region chrX:60000000-80000000 &\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
